File: boost_python/yay.py
#!/usr/bin/env python
"""Some usage examples in python"""
import libyay
from pprint import pprint as pp

pp(vars(libyay))

a_string = libyay.yay()
print a_string

hello = libyay.World()
hello.set('This is a struct!!')
print hello.greet()

hello_ = libyay.WorldClass()
print hello_.greet()
hello_.set('yay for C++ and Python!')
print hello_.greet()

File: boost_python/yay.cxx
#ifndef _YAY
#define _YAY
// c++ source
char const* yay() { return "Yay!"; }

struct World{
    void set(std::string msg) { this->msg = msg; }
    std::string greet() { return msg; }
    std::string msg;
};

class WorldClass {
public:
    WorldClass() { this->msg = "Yay for classes!"; }
    ~WorldClass() { std::cout << "Goodbye!\n"; }
    void set(std::string msg) { this->msg = msg; }
    std::string greet() { return msg; }
// private:
    std::string msg;
};

#endif

// make a python interface
#include <boost/python.hpp>
BOOST_PYTHON_MODULE(libyay) // this name is what you will import
{   using namespace boost::python;
    
    def("yay", yay);

    class_<World>("World")
        .def("greet", &World::greet)
        .def("set", &World::set)
    ;
    
    class_<WorldClass>("WorldClass")
        .def("greet", &WorldClass::greet)
        .def("set", &WorldClass::set)
    ;
}
